### PR TITLE
MM-39058 - invite people to join team and channels

### PR DESCRIPTION
**Pull request by pvev**
#### Summary
This PR adds the option to invite members directly to channels. This creates a new set of functions to cover the functionality to being able to invite people to join a team and a channel directly. Also modifies the emails templates to reflect these invitations.

#### Related Pull Requests
Webapp: https://github.com/mattermost/mattermost-webapp/pull/10083

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-42543

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Review by neallred — COMMENTED** [review]
Looking good! A few nits, but nothing major.

File: api4/team.go
Comment (JulienTant) [inline api4/team.go:1278]: if you instantiate here as `&model.MemberInvite{}`, you could drop the `&` from the next line but also from function call `App.InviteNewUsersToTeamGracefully`

File: app/email/email.go
Comment (JulienTant) [inline app/email/email.go:635]: we could use the keys `invite_team_and_channels_subject` and `invite_team_and_channel_subject` (no s at channel, or change with `single_channel`) to avoid the a/b suffixes and have more explicit names

File: app/email/email.go
Comment (JulienTant) [reply app/email/email.go]: ^ 👍🏻

File: app/team.go
Comment (JulienTant) [inline app/team.go]: `teamID string, senderId string`; I don't know if you added the first string on purpose, but it's not required in go to declare a parameter  type if the following one has the same type.

File: app/team.go
Comment (JulienTant) [inline app/team.go]: having **one** go routine immediately followed by a blocking channel read seems unnecessary.

I'm pretty sure this will produce the same result:, and might even be faster
```suggestion
		channels, err := a.Srv().Store.Channel().GetChannelsByIds(memberInvite.Channels, false)
		err != nil {
			return nil, nil, nil, model.NewAppError("prepareInviteNewUsersToTeam", "app.channel.get_channels_by_ids.app_error", nil, result.NErr.Error(), http.StatusInternalServerError)
		}
```

**Review by JulienTant — COMMENTED** [review]
Good job @pvev! I left a few remarks and nits

File: model/member_invite.go
Comment (pvev) [reply model/member_invite.go]: changed!

File: app/email/email.go
Comment (pvev) [reply app/email/email.go]: done!

File: app/team.go
Comment (pvev) [reply app/team.go]: great advices! thanks.

File: app/email/email.go
Comment (pvev) [reply app/email/email.go:635]: make sense, thanks!

File: api4/team.go
Comment (pvev) [reply api4/team.go:1278]: ok, thanks.

File: app/team.go
Comment (pvev) [reply app/team.go]: ok

File: app/team.go
Comment (pvev) [reply app/team.go]: going to give it a try.

File: app/team.go
Comment (pvev) [reply app/team.go]: thanks for the comment! solved

File: app/email/email.go
Comment (pvev) [reply app/email/email.go:635]: nice catch! solved

File: api4/team.go
Comment (pvev) [reply api4/team.go:1278]: done

File: api4/team.go
Comment (marianunez) [inline api4/team.go:1293]: The `isValid` seems to be checking if all emails are valid and failing if any are. I believe this would break the send gracefuly functionality given that it will not send any if there is one that is not valid. Before it would send all that it could and return the error of which ones failed.

File: api4/team.go
Comment (marianunez) [inline api4/team.go:1297]: Additionally, it seems that this is included in the valid check above so we should converge in checking in one place.

File: api4/team.go
Comment (marianunez) [inline api4/team.go:1279]: Would this be considered a breaking change to the api?

**Review by marianunez — CHANGES_REQUESTED** [review]
Great work @pvev! Some comments regarding the new model for the api request and verification below 👇

File: api4/team.go
Comment (pvev) [reply api4/team.go:1279]: yes, because the api call was expecting before an array of strings and now expect a struct defining at least the emails property which is that array of strings. I think is a better idea to support both data structures so we don't break the api current behavior.

File: api4/team.go
Comment (pvev) [reply api4/team.go:1293]: good catch! thanks Maria, with this validation the "gracefully" behavior is broken.

**Review by JulienTant — COMMENTED** [review]
Thanks for the fixes. I left a comment around the unmarshal.

File: api4/team.go
Comment (pvev) [reply api4/team.go:1280]: the idea with this function https://github.com/mattermost/mattermost-server/pull/19849/commits/684c987015de73a4b84fbe7dae2c2fdffe8d6cdc#diff-0b9cc7e6646187be630de194b83272a908f89db4fbe824b5c336fca57e938840R34 is to cover both scenarios, when you send a plain array of emails ["email1@email", "email2@email"] or a data structure that matches with the structure of memberInvite struct {emails: ["...], channels: ["..."]}.

File: api4/team.go
Comment (pvev) [reply api4/team.go:1284]: ok!

Comment (pvev) [reply api4/team.go:1284]: what you are suggesting is what it does with the overwrite of the Unmarshal method in the inviteMembers struct. With that Unmarshal function we are able to process both inputs, the array of emails and the memberInvite struct. In case of array of emails, we turn it into a memberInvite struct. If you check the webapp code, the current members invite modal, is still sending the array of emails and it is being parsed correctly, the same as when you open the add members channel and try to invite a member, it will send the memberInvite structure, and both work fine.

Captura de pantalla 2022-03-29 a las 12 55 17
Captura de pantalla 2022-03-29 a las 12 55 46
Captura de pantalla 2022-03-29 a las 12 56 00

File: api4/team_local.go
Comment (pvev) [reply api4/team_local.go:85]: got it

File: model/member_invite.go
Comment (neallred) [reply model/member_invite.go:34]: ahh, yep, missed that! This also addresses my other comment about parsing both ways to limit API breakage. Nicely done!

File: api4/team_test.go
Comment (marianunez) [inline api4/team_test.go:2986]: Could we add another test to make sure both inputs are still valid?

File: model/client4.go
Comment (marianunez) [inline model/client4.go:2614]: We should have another option to invite to team with email + channel to make sure we are not breaking compatabiility

File: model/member_invite.go
Comment (marianunez) [inline model/member_invite.go:13]: Should we rename this to `ChannelIds` so that it's clear that is not an array of channels?

**Comment by marianunez** [comment]
Adding @AGMETEOR for visibility on changes related to email invites reminders

File: model/client4.go
Comment (pvev) [reply model/client4.go:2614]: compatibility is working. The current member email invite flow keeps working the same, sending a plain email-strings array. Only when the member invite modal is opened from the add member to channel modal (by clicking the 'invite them to the team' new link), the data structure changes to the new memberInvite model structure model.

File: api4/team_test.go
Comment (pvev) [reply api4/team_test.go:2986]: the test is already there https://github.com/mattermost/mattermost-server/pull/19849/files#diff-f157eb201938e9e26eff6958772bfa684b18152b6697e7cc94fb2261ef4f3003R1326

File: model/member_invite.go
Comment (JulienTant) [reply model/member_invite.go:34]: 👏🏻

**Review by JulienTant — APPROVED** [review]
LGTM @pvev, smart move to use a custom unmarshaller to handle both types of input.
Not sure how to reflect those two options on the API documentation tho 🤔

File: api4/team_test.go
Comment (marianunez) [reply api4/team_test.go:2986]: In that test you are passing also a MemberInvite model with only emails. We should test that the api endpoint can receive just an array of strings like it did before

File: model/client4.go
Comment (marianunez) [inline model/client4.go:2654]: Same here, it should still work without this change

**Review by marianunez — APPROVED** [review]
Great work @pvev 💯

**Comment by pvev** [comment]
maybe @AGMETEOR can take a look at this after merge. @marianunez ^^

**Comment by marianunez** [comment]
> maybe @AGMETEOR can take a look at this after merge. @marianunez ^^

Sure, no need to hold the PR with all the approvals complete. 👍

**Pull request merged**
